tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,455 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// See docs in ../ops/linalg_ops.cc.
+// TODO(shamanDevel): Enable complex inputs. This will require additional tests
+//                    and OP_REQUIRES.
+#if GOOGLE_CUDA
+#define EIGEN_USE_GPU
+
+#include <algorithm>
+#include <vector>
+
+#include "tensorflow/core/framework/kernel_def_builder.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/kernels/linalg_ops_common.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/kernels/cuda_solvers.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+// I need to transpose V afterwards
+#include "tensorflow/core/kernels/transpose_functor.h"
+
+// Logging
+#include <stdio.h>
+
+namespace tensorflow {
+
+static const char kErrMsg[] =
+    "Singular Value Decomposition was not successful. The input might not be "
+    "valid.";
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace {
+  template<class Scalar>
+  __global__ void ComputeV1Kernel(CudaLaunchConfig config,
+      int64 m, const Scalar* M, const Scalar* U, const Scalar* S, Scalar* V)
+  {
+    CUDA_1D_KERNEL_LOOP(i, config.virtual_thread_count) {
+      Scalar v = M[i] * U[m*i] * S[0];
+      //printf("i=%d, v=%f\n", (int) i, (float) v);
+      CudaAtomicAdd(V, v);
+    }
+  }
+
+  template<class Scalar>
+  __global__ void ComputeV2Kernel(CudaLaunchConfig config, Scalar* V)
+  {
+    CUDA_1D_KERNEL_LOOP(i, config.virtual_thread_count) {
+      //printf("V'[%d] = %f\n", (int)i, (float)V[i]);
+      V[i] = (Scalar) copysign(1.0, V[i]);
+    }
+  }
+}
+
+// Scalar: The input scalar type (can be complex)
+// SScalar: The output type for the singular value,
+//   same as Scalar if real, or the real version if Scalar is complex
+template <class Scalar, class SScalar>
```

Comment:
Instead of adding another template parameters, you can get the real type corresponding to a complex (or real) type like this:
```
using RealScalar = typename Eigen::NumTraits<Scalar>::Real;
```

Reply:
Oh, I was looking for exactly that but couldn't find it. Thanks.